### PR TITLE
Allow deserialization of missing objects properties into Option<>

### PR DESCRIPTION
**Pull request by hansl**
Currently the TryFromJs derive macro requires that all properties be defined in the object, whether they are declared as Option<> or not. This commit makes it so if the field is not found in the JsObject, we try to deserialize undefined.

----

It changes the following:

- When using the macro to deserialize a struct, if a property descriptor is not found for a field, use `JsValue::undefined()` to deserialize instead.
- Add unit tests for the above change.

This is more in line with JavaScript's expected behaviour. It also allows structure to have partial defined fields.

### CODE REVIEW

**Comment by hansl** [comment]
I decided to make this PR tentatively as a start for the discussion. I was not sure if the current behaviour was discussed before (I could not find an discussion, issues or PRs that mentioned it).

I think this PR makes sense in the context where people are interfacing rust code with JavaScript calls, as it removes a lot of unnecessary JS when calling into rust synthetic modules.

**Comment by jedel1043** [comment]
I'm pretty sure the current behaviour is not intentional, but I'm pinging @Razican in case I'm wrong.

**Comment by hansl** [comment]
@jedel1043 That's what I was hoping for.

Do you know why CI is not happy? I think it detects `boa_engine` in the `dev-dependencies` but doesn't realize it _should_ be unused in the crate, only in its tests (and it is used in `/core/macros/tests/try_from_js.rs`).

**Comment by Razican** [comment]
This behavior was not defined at the beginning. What should we do when a field doesn't exist? What if the object has more fields?

serde has attributes for these cases. I'm happy to go with whatever serde does :)

**Comment by hansl** [comment]
I cannot reproduce the CI doc test failures locally. And it seems unrelated to my changes.

**Comment by hansl** [comment]
@Razican For de-serialization, `serde` sets `Option<>` as optional and missing, so I surmise this PR should go in at a first step to fixing this. If you want to fully go down the `serde` route I would suggest implementing `TryFromJs` for `serde::Deserialize` and remove the derive macro entirely. That would be outside of the scope of this PR though.

**Comment by jedel1043** [comment]
Will take a look to the CI errors.

**Comment by jedel1043** [comment]
Fixed. The circular dependency between `boa_engine` and `boa_macros` was causing some linking problems, that's why we had to extract the `boa_macros` tests to an external crate.

**Comment by hansl** [comment]
Oh thank you. Somehow I missed that. This is ready for review as is then. Personally I'd rather this goes in and then work separately on serde-ifying TryFromJs.

**Review by HalidOdat — APPROVED** [review]
Nice work! Looks good to me! :)

File: core/macros/src/lib.rs [hunk 1]
```diff
@@ -418,30 +418,19 @@ fn generate_conversion(fields: FieldsNamed) -> Result<proc_macro2::TokenStream,
             .map_err(|err| vec![err])?;
         }
 
+        final_fields.push(quote! {
+            let #name = match props.get(&::boa_engine::js_string!(#name_str).into()) {
+                Some(pd) => pd.value().ok_or_else(|| ::boa_engine::JsError::from(
+                        boa_engine::JsNativeError::typ().with_message(#error_str)
```
Comment (jedel1043) [inline core/macros/src/lib.rs:424]:
```suggestion
                        ::boa_engine::JsNativeError::typ().with_message(#error_str)
```

**Review by jedel1043 — COMMENTED** [review]
Everything looks good! Just a missing double colon.

File: core/macros/src/lib.rs (hunk 1 above)
Comment (hansl) [reply core/macros/src/lib.rs]: Done.

**Review by jedel1043 — APPROVED** [review]
Thank you for the contribution!

**Pull request merged**